File: 974.subarray-sums-divisible-by-k.py
import collections
from typing import List

#
# @lc app=leetcode id=974 lang=python3
#
# [974] Subarray Sums Divisible by K
# python的 mod % 会自动把负数转成正数, -1 % 5 = 4. 所以比 java 的少操作一步.

# @lc code=start
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        modCount = collections.Counter([0])
        res = 0
        prefix = 0
        
        for num in nums:
            prefix = (prefix + num) % k
            res += modCount[prefix]
            modCount[prefix] += 1
        
        return res
        
# @lc code=end
so = Solution();
so.subarraysDivByK([4,5,0,-2,-3,1], 5)

File: 974.subarray-sums-divisible-by-k.java
import java.util.Map;

/*
 * @lc app=leetcode id=974 lang=java
 *
 * [974] Subarray Sums Divisible by K
 * 
 * find some (presum[i]-presum[j])%5 == 0, equals 
 * find some presum[i]%5 == presum[j]%5
 * 
 * 另外要注意和为负数情况. python 里面会自动转正数, java 不会.
 */

// @lc code=start
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        int presum = 0;
        Map<Integer, Integer> map = new HashMap<>();
        int res = 0;
        map.put(0, 1);

        for (int i = 0; i < nums.length; i++) {
            presum = (presum + nums[i]) % k;
            if (presum < 0) {
                presum += k;
            }
            if (map.containsKey(presum)) {
                res += map.get(presum);
            }

            map.put(presum, map.getOrDefault(presum, 0) + 1);
        }
        return res;
    }
}
// @lc code=end
